src/linker/Linker.Dataflow/FlowAnnotations.cs

```diff
@@ -0,0 +1,340 @@
+							// TODO: warn: duplicate annotation. not propagating.
+						} else {
+							int offset = setMethod.HasThis && !setMethod.ExplicitThis ? 1 : 0;
+							if (setMethod.Parameters.Count > 0) {
+								DynamicallyAccessedMemberKinds [] paramAnnotations = new DynamicallyAccessedMemberKinds [setMethod.Parameters.Count + offset];
+								paramAnnotations [offset] = annotation;
+								annotatedMethods.Add (new MethodAnnotations (setMethod, paramAnnotations, 0));
+							}
+						}
+					}
+
+					FieldDefinition backingFieldFromGetter = null;
+
+					// Propagate the annotation to the setter method
+					MethodDefinition getMethod = property.GetMethod;
+					if (getMethod != null) {
+
+						if (ScanMethodBodyForFieldAccess (getMethod.Body, write: false, out backingFieldFromGetter)) {
+							// TODO: warn we couldn't find a unique backing field
+						}
+
+						if (annotatedMethods.Any (a => a.Method == getMethod)) {
+							// TODO: warn: duplicate annotation. not propagating.
+						} else {
+							annotatedMethods.Add (new MethodAnnotations (getMethod, null, annotation));
+						}
+					}
+
+					FieldDefinition backingField;
+					if (backingFieldFromGetter != null && backingFieldFromSetter != null &&
+						backingFieldFromGetter != backingFieldFromSetter) {
+						// TODO: warn we couldn't find a unique backing field
+						backingField = null;
+					} else {
+						backingField = backingFieldFromGetter ?? backingFieldFromSetter;
+					}
+
+					if (backingField != null) {
+						if (annotatedFields.Any (a => a.Field == backingField)) {
+							// TODO: warn about duplicate annotations
+						} else {
+							annotatedFields.Add (new FieldAnnotation (backingField, annotation));
+						}
+					}
+				}
+			}
+
+			return new TypeAnnotations (annotatedMethods.ToArray(), annotatedFields.ToArray());
+		}
+
+		private bool ScanMethodBodyForFieldAccess (MethodBody body, bool write, out FieldDefinition found)
+		{
+			// Tries to find the backing field for a property getter/setter.
+			// Returns true if this is a method body that we can unambiguous analyze.
+			// The found field could still be null if there's no backing store.
+
+			// TODO: could restrict this to compiler-generated fields as well so that
+			// it only works for auto properties, but maybe that's too restrictive.
+
+			found = null;
+
+			foreach (Instruction instruction in body.Instructions) {
+				switch (instruction.OpCode.Code) {
+					case Code.Ldsfld when !write:
+					case Code.Ldfld when !write:
+					case Code.Stsfld when write:
+					case Code.Stfld when write:
+
+						FieldDefinition field = (instruction.Operand as FieldReference)?.Resolve ();
+						if (field != null && field.IsStatic == body.Method.IsStatic) {
+							if (found != null) {
+								// This writes/reads multiple fields - can't guess which one is the backing store.
+								found = null;
+								return false;
+							}
+
+							found = field;
+						}
```

Comment:
Should we fail if the field can't be resolved or it's not static as expected? Because if the getter/setter has two field accesses, one which passes this check and one which doesn't we will still treat it as "yeah, we completely understand this one" - which is not true. I would rather this to be very careful and only return non-null when it's absolutely sure it got it right.

Reply:
If we want to be absolutely sure, we need to run the flow analysis on this.

E.g.

```csharp
[DynamicallyAccessedMembers(Constructors)]
Type BaseType => _isValueType ? typeof(ValueType) : typeof(Object);
```

This would currently incorrectly propagate the annotation to the `_isValueType` field. We could check that the type of the field matches the type of the property but even that one is not certain.

That's why I left the `TODO: could restrict this to compiler-generated fields as well so that it only works for auto properties, but maybe that's too restrictive.` - maybe we should do that...

Reply:
Hmm - yeah - I'm starting to like that more - it is restrictive, but on the other hand it's correct - which wins it for me.